packages/sdk/src/search/useSearchState.ts

```diff
@@ -0,0 +1,226 @@
+/* eslint-disable no-case-declarations */
+import { useReducer } from 'react'
+
+import { SDKError } from '../utils/error'
+
+export type SearchSort =
+  | 'price_desc'
+  | 'price_asc'
+  | 'orders_desc'
+  | 'name_desc'
+  | 'name_asc'
+  | 'release_desc'
+  | 'discount_desc'
+  | 'score_desc'
+
+export interface Facet {
+  key: string
+  value: string
+}
+
+export interface State {
+  /** @description search sorting criteria */
+  sort: SearchSort
+  /**
+   * @description selected facets
+   * */
+  selectedFacets: Facet[]
+  /** @description full text term */
+  term: string | null
+  /**
+   * @description the base path url for the search context
+   * */
+  base: string
+  /**
+   * @description current pagination cursor
+   */
+  page: number
+}
+
+const sortKeys = new Set<SearchSort>([
+  'price_desc',
+  'price_asc',
+  'orders_desc',
+  'name_desc',
+  'name_asc',
+  'release_desc',
+  'discount_desc',
+  'score_desc',
+])
+
+export const initialize = ({
+  sort = 'score_desc',
+  selectedFacets = [],
+  term = null,
+  base = '/',
+  page = 0,
+}: Partial<State> | undefined = {}) => ({
+  sort,
+  selectedFacets,
+  term,
+  base: base.endsWith('/') ? base : `${base}/`,
+  page,
+})
+
+const isSearchSort = (x: string): x is SearchSort => sortKeys.has(x as any)
+
+type Action =
+  | {
+      type: 'setSort'
+      payload: SearchSort
+    }
+  | {
+      type: 'setTerm'
+      payload: string | null
+    }
+  | {
+      type: 'setPage'
+      payload: number
+    }
+  | {
+      type: 'setFacet'
+      payload: { facet: Facet; unique: boolean }
+    }
+  | {
+      type: 'removeFacet'
+      payload: Facet
+    }
+  | {
+      type: 'toggleFacet'
+      payload: Facet
+    }
+  | {
+      type: 'toggleFacets'
+      payload: Facet[]
+    }
+
+const removeFacet = (state: State, facet: Facet): State => {
+  const { value } = facet
+
+  const index = state.selectedFacets.findIndex((x) => x.value === value)
+
+  if (index < 0) {
+    throw new SDKError(`Cannot remove ${value} from search params`)
+  }
+
+  // We can't allow removing the first facet, otherwise we would loose
+  // the navigation context
+  //
+  // TODO: Remove returning the base selected facets to the frontend since
+  // we won't be unselecting it anyways
+  return {
+    ...state,
+    selectedFacets: state.selectedFacets.filter(
+      (_, it) => it === 0 || it !== index
+    ),
+  }
+}
+
+export const setFacet = (
+  state: State,
+  facet: Facet,
+  unique: boolean
+): State => {
+  if (unique === true) {
+    const index = state.selectedFacets.findIndex((f) => f.key === facet.key)
```

Comment:
doesn't it make it more difficult to read? `f` is already local because of JavaScript's scope definition and is only being used for a simple findIndex comparison. Adding meaning to variables is good when you need it to reason about the code. When you have an auxiliary variable that doesn't impact on what the code is doing, adding meaning to them can make the code more difficult to read, since now I need to pay attention to that variable too. What do you think. about this point of view?

Reply:
Hmmm, i partially agree with 😅 . This case is pretty easy to understand, but others aren't easy to understand fast. So, I will open an exception for this case, haha.